jobs/middlewares.go

```diff
@@ -0,0 +1,178 @@
+package jobs
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log/slog"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/checkmarble/marble-backend/utils"
+	"github.com/getsentry/sentry-go"
+	"github.com/riverqueue/river/rivertype"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	sentryErrorGroupingTime = 30 * time.Second
+	sdkIdentifier           = "sentry.go.river.marble"
+)
+
+// Logger middleware
+
+type LoggerMiddleware struct {
+	l              *slog.Logger
+	errorCount     map[string]int
+	errorCountLock *sync.Mutex
+}
+
+func (m LoggerMiddleware) Work(ctx context.Context, job *rivertype.JobRow, doInner func(context.Context) error) error {
+	logger := m.l.With(
+		"job_id", job.ID,
+		"job_kind", job.Kind,
+		"job_attempt", job.Attempt,
+		"last_attempted_at", job.AttemptedAt,
+		"created_at", job.CreatedAt,
+		"queue", job.Queue,
+		"priority", job.Priority,
+	)
+	start := time.Now()
+	logger.InfoContext(ctx, fmt.Sprintf("Starting %s job n°%d - attempt %d", job.Kind, job.ID, job.Attempt))
+
+	ctx = utils.StoreLoggerInContext(ctx, logger)
+	err := doInner(ctx)
+	if err != nil && isSnoozeError(err) {
+		logger.InfoContext(ctx, fmt.Sprintf("%s job n°%d snoozed after %s", job.Kind, job.ID, time.Since(start)))
+		return err
+	} else if err != nil {
+		logger.ErrorContext(ctx, fmt.Sprintf("%s job n°%d failed after %s", job.Kind, job.ID, time.Since(start)))
+		m.aggregateAndLogError(ctx, job, err)
+		return err
+	}
+
+	logger.InfoContext(ctx, fmt.Sprintf("%s job n°%d succeeded after %s", job.Kind, job.ID, time.Since(start)))
+	return nil
+}
+
+func (m LoggerMiddleware) aggregateAndLogError(ctx context.Context, job *rivertype.JobRow, err error) {
+	m.errorCountLock.Lock()
+	defer m.errorCountLock.Unlock()
+
+	errorKey := fmt.Sprintf("%s:%s", job.Kind, err.Error())
+	m.errorCount[errorKey]++
+
+	if m.errorCount[errorKey] == 1 {
+		go func() {
+			time.Sleep(sentryErrorGroupingTime)
```

Comment:
Try to do some level of aggregation before sending errors to sentry, if errors occur on batch jobs. I don't know that the current value of 30sec is enough (in fact, it probably isn't) but let's see.